src/compiler/moduleNameResolver.ts

```diff
@@ -150,31 +150,41 @@ namespace ts {
         }
 
         if (currentDirectory !== undefined) {
-            return getDefaultTypeRoots(currentDirectory, host);
+            return getDefaultTypeRoots(currentDirectory, host, options.searchScopes);
         }
     }
 
     /**
-     * Returns the path to every node_modules/@types directory from some ancestor directory.
+     * Returns the path to every type search scope directory from some ancestor directory.

```

Comment:
i am actually not sure what the intent of this feature is now.. [I thought](https://github.com/Microsoft/TypeScript/issues/23952#issuecomment-387573894) this was manly for .js file intellisense in VSCode, and you wanted the server to download types from `@Company\types` instead of `@types`.. 

if that is the case, then no changes are needed here, changes should be done to the typinging installer, and to the tsserver module resolver.

Reply:
If this is not changed, then how will types installed via `npm install --save-dev @company-types/companyPackageName` work in the same manner as `npm install --save-dev @types/packageName`?

Reply:
so why not use typeRoots then?

Reply:
Because that would break the default type acquisition. I want to *supplement* it, not rip it to shreds! From the docs, emphasis mine:

By default all *visible* "`@types`" packages are included in your compilation.
Packages in `node_modules/@types` of any enclosing folder are considered *visible*;
specifically, that means packages within `./node_modules/@types/`,  `../node_modules/@types/`, `../../node_modules/@types/`, and so on.

***If `typeRoots` is specified, *only* packages under `typeRoots` will be included.
For example:***

```json
{
   "compilerOptions": {
       "typeRoots" : ["./typings"]
   }
}
```

This config file will include *all* packages under `./typings`, and no packages from `./node_modules/@types`.

Reply:
I repeat:

# If typeRoots is specified, only packages under typeRoots will be included.
For example:

# This config file will include all packages under ./typings, and no packages from ./node_modules/@types.

Reply:
The `if` block on lines 140-142 are why I am implementing a new option and not using `typeRoots`. If I set `typeRoots`, I opt out of the standard behavior for finding typings and completions in the editor experience.

https://github.com/Microsoft/TypeScript/blob/77168e5d500269ab6c044d688b53255077fa9201/src/compiler/moduleNameResolver.ts#L139-L155

Reply:
**I do not want to opt out of the standard behavior of finding typings, I want to add to it and allow the users to specify additional scopes that are treated just like `@types`.**

Reply:
I am sorry I am so frustrated but I feel like I have been as clear as I can be about the use case, setting `typeRoots` short-circuits `getEffectiveTypeRoots`, which changes the behavior of the compiler and editor functionality. It changes the behavior in the module name resolver, resolution cache, language server module name finding and code completions. Search for every place `getEffectiveTypeRoots` is called. All of those call sites are short-circuited to return the fixed list the user provided instead of the recursive searching done by default.

Reply:
> fixed list the user provided instead of the recursive searching done by default.

Ah. So you're saying you rely on the directory traversal to find `@types` (or corporate `@companyname-types`) folders rather than relying, eg, on a single top-level `node_modules/@types` and `node_modules/@companyname-types` which you could easily include in your `typeRoots` list. Got it.

Reply:
YES! Thank you! I am so sorry it took so many words to express this.

Reply:
No worries; we were wondering about the utility of the same when we initially implemented `typeRoots` a few years ago.

Reply:
So how many node_modules do you have in your project typically? I ask cause if they are one or two, typeRoots should do the trick.

Also sorry if I am asking questions, but u originally said ATA and that is a totally different feature.

Reply:
I want it to work just like `@types` does, full stop. I suspect this will take two sets of changes, the first here to make the compiler treat `searchScopes` as an alternate source for the string "@types", and I'm not quite sure about how to implement ATA.

I think for ATA to work correctly, I will need to modify the `npm` command the server runs?

Reply:
Can you tell us more about your use case? how does this come into play in your organization? and why typeRoots is not sufficient for your use case?

Reply:
The use case is an internal NPM-compatible package registry, and wanting to put typings in it to avoid placing them in upstream projects, and still support automatic type acquisition.

I don't like typeroots because (a.) it changes the default type resolution algorithm `getEffectiveTypeRoots`, and I don't know what packages might depend on that and there must be *some* reason for walking the spine of the directory in the default behavior, and I don't want to change that behavior, and (b.) it doesn't enable automatic type acquisition.

Reply:
To be clear:

1. Setting `typeRoots: ["node_modules/@types"]` is not a no-op. I don't know why it's not a no-op, I couldn't find a rationale for the `@types` scope behavior, but it *is*. That's just how it is.

2. Setting `typeRoots: ["node_modules/@types"]` could subtly break compilation because of that. I don't know under what situations that would happen - I suspect in monorepo (`lerna` or `yarn` workspaces) projects it could cause `./node_modules/@types` installed at the root not to resolve in child projects. Whatever the case, the fact is that setting this changes behavior and could cause projects to not compile for failure to resolve @types. If that isn't true, then there's no reason for `getEffectiveTypeRoots` to short circuit.

3. Setting `typeRoots: ["node_modules/@companyTypes", "node_modules/@types"]` does not enable automatic type acquisition for the scope `@companyTypes`, and could subtly break compilation as above.